userspace/falco/app_actions/process_events.cpp

```diff
@@ -0,0 +1,262 @@
+					m_state->outputs->handle_msg(now, falco_common::PRIORITY_DEBUG, msg, rule, o);
+					// Reset the timeouts counter, Falco alerted
+					timeouts_since_last_success_or_msg = 0;
+				}
+			}
+
+			continue;
+		}
+		else if(rc == SCAP_EOF)
+		{
+			break;
+		}
+		else if(rc != SCAP_SUCCESS)
+		{
+			//
+			// Event read error.
+			//
+			cerr << "rc = " << rc << endl;
+			throw sinsp_exception(m_state->inspector->getlasterr().c_str());
+		}
+
+		// Reset the timeouts counter, Falco successfully got an event to process
+		timeouts_since_last_success_or_msg = 0;
+		if(duration_start == 0)
+		{
+			duration_start = ev->get_ts();
+		}
+		else if(duration_to_tot_ns > 0)
+		{
+			if(ev->get_ts() - duration_start >= duration_to_tot_ns)
+			{
+				break;
+			}
+		}
+
+		if(!sdropmgr.process_event(m_state->inspector, ev))
+		{
+			result.success = false;
+			result.errstr = "";
+			result.proceed = false;
+			break;
+		}
+
+		if(!ev->simple_consumer_consider() && !m_options.all_events)
+		{
+			continue;
+		}
+
+		// As the inspector has no filter at its level, all
+		// events are returned here. Pass them to the falco
+		// engine, which will match the event against the set
+		// of rules. If a match is found, pass the event to
+		// the outputs.
+		unique_ptr<falco_engine::rule_result> res = m_state->engine->process_event(m_state->event_source_idx, ev);
+		if(res)
+		{
+			m_state->outputs->handle_event(res->evt, res->rule, res->source, res->priority_num, res->format, res->tags);
+		}
+
+		num_evts++;
+	}
+
+	return num_evts;
+}
+
+application::run_result application::process_events()
+{
+	syscall_evt_drop_mgr sdropmgr;
+	// Used for stats
+	double duration;
+	scap_stats cstats;
+
+	run_result ret;
+
+	duration = ((double)clock()) / CLOCKS_PER_SEC;
+
+	if(!m_options.trace_filename.empty() && !m_state->trace_is_scap)
+	{
+#ifndef MINIMAL_BUILD
+		read_k8s_audit_trace_file(m_options.trace_filename);
+#endif
+	}
+	else
+	{
+		uint64_t num_evts;
+
+		num_evts = do_inspect(sdropmgr,
+				      uint64_t(m_options.duration_to_tot*ONE_SECOND_IN_NS),
+				      ret);
+
+		duration = ((double)clock()) / CLOCKS_PER_SEC - duration;
+
+		m_state->inspector->get_capture_stats(&cstats);
+
+		if(m_options.verbose)
+		{
+			fprintf(stderr, "Driver Events:%" PRIu64 "\nDriver Drops:%" PRIu64 "\n",
+				cstats.n_evts,
+				cstats.n_drops);
+
+			fprintf(stderr, "Elapsed time: %.3lf, Captured Events: %" PRIu64 ", %.2lf eps\n",
+				duration,
+				num_evts,
+				num_evts / duration);
+		}
+
+	}
+
+	// Honor -M also when using a trace file.
+	// Since inspection stops as soon as all events have been consumed
+	// just await the given duration is reached, if needed.
+	if(!m_options.trace_filename.empty() && m_options.duration_to_tot>0)
+	{
+		std::this_thread::sleep_for(std::chrono::seconds(m_options.duration_to_tot));
+	}
+
+	m_state->inspector->close();
```

Comment:
Shouldn't this be part of the teardown actions?

Reply:
Yeah I'll make it a close_inspector step.